runtime/src/main/java/org/corfudb/runtime/Restore.java

```diff
@@ -0,0 +1,165 @@
+    private List<UUID> streamIDs;
+
+    // The Corfu Store associated with the runtime
+    private CorfuStore corfuStore;
+
+    /**
+     * Unpacked files from backup tar file are stored under RESTORE_TEMP_DIR. They are deleted after restore finishes.
+     */
+    private static final String RESTORE_TEMP_DIR_PREFIX = "corfu_restore_";
+
+    /***
+     * @param filePath      - the path of backup tar file
+     * @param runtime       - the runtime which is performing the restore
+     * @throws IOException when failed to create the temp directory
+     */
+    public Restore(String filePath, CorfuRuntime runtime) throws IOException {
+        this.filePath = filePath;
+        this.restoreTempDirPath = Files.createTempDirectory(RESTORE_TEMP_DIR_PREFIX).toString();
+        this.streamIDs = new ArrayList<>();
+        this.corfuStore = new CorfuStore(runtime);
+    }
+
+    /**
+     * Start the restore process
+     *
+     * @throws IOException
+     */
+    public void start() throws IOException {
+        try {
+            openTarFile();
+            verify();
+            restore();
+        } catch (IOException e) {
+            log.error("failed to restore from backup tar file {}", filePath);
+            throw e;
+        } finally {
+            cleanup();
+        }
+        log.debug("restore completed");
+    }
+
+    private void restore() throws IOException {
+        for (UUID streamId : streamIDs) {
+            String fileName = restoreTempDirPath + File.separator + streamId;
+            try {
+                restoreTable(fileName, streamId);
+            } catch (IOException e) {
+                log.error("failed to restore table {} from temp file {}", streamId, fileName);
+                throw e;
+            }
+        }
+    }
+
+    /**
+     * Restore a single table
+     *
+     * @param fileName   - the name of the temp backup file
+     * @param streamId   - the stream ID of the table which is to be restored
+     * @throws IOException
+     */
+    private void restoreTable(String fileName, UUID streamId) throws IOException {
+        try (FileInputStream fileInput = new FileInputStream(fileName)) {
+            // Clear table before restore
+            TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE);
+            SMREntry entry = new SMREntry("clear", new Array[0], Serializers.PRIMITIVE);
+            txn.logUpdate(streamId, entry);
+
+            long numEntries = 0;
+            while (fileInput.available()> 0) {
+                OpaqueEntry opaqueEntry = OpaqueEntry.read(fileInput);
+                List<SMREntry> smrEntries = opaqueEntry.getEntries().get(streamId);
+                if (smrEntries == null || smrEntries.isEmpty()) {
+                    continue;
+                }
+
+                txn.logUpdate(streamId, smrEntries);
+                numEntries++;
+            }
+            txn.commit();
+            log.debug("write uuid {} with {} numEntries", streamId, numEntries);
```

Comment:
can we add an explicit log.info stating: 'Completed restore of table {}' ... (something in that line)

Reply:
Changed to `log.info("completed restore of table {} with {} numEntries", streamId, numEntries);`. Thanks.